reframe/core/pipeline.py

```diff
@@ -799,7 +749,13 @@ def __new__(cls, *args, _rfm_use_params=False, **kwargs):
                         os.path.dirname(inspect.getfile(cls))
                     )
 
-        obj._rfm_init(name, prefix)
+        # Apply the hooks only once (not when deep-copying)
+        if not hasattr(cls, '_rfm_first_init'):
```

Comment:
> I think the decoration of the pipeline stages should happen in the pipeline

What do you mean by this? In the frontend that executes the tests?

Reply:
> (Reading this _rfm_first_init immediately triggered horror memories of Fortran 77 code 😱😂)

Practically, this is not too different from the `_rfm_use_params` argument 😬

Reply:
Indeed, it guards also against re-decorating an inherited function. I will record the attached hooks and do not re-attach them, so that we can avoid this, and also be able to do the sanity check that you propose.

Reply:
This guard, apart from "old-fashioned", is also wrong, because although it will not reattach existing hooks, it will fail to attach new hooks of the derived class. I'm working on alternative designs.